### ISSUE TITLE
analysis.neoRiemannian.LRP_combinations has different behaviors when instanciating the chord with pitch class numbers or symbols.

### ISSUE THREAD

**Issue by ax-le** [cuthbertLab/music21#928]
**music21 version**

`6.7.1`   

**Problem summary**
Instanciating a chord with its pitch class numbers or with its pitch class symbols leads to different behaviors, and the former may raise an Exception in the analysis.neoRiemannian.LRP_combinations function. It comes from the Chord instanciation itself.

**Steps to reproduce**
Instanciating the chord with its pitch class symbols works fine, like in the example below:
```
b_maj = music21.chord.Chord('B D# F#')
lp_b_maj = music21.analysis.neoRiemannian.LRP_combinations(b_maj, "LP", raiseException=False)
```

However, instanciating the same chord with its pitch class numbers raises an LRPException:
```
b_maj = music21.chord.Chord([11, 3, 6])
lp_b_maj = music21.analysis.neoRiemannian.LRP_combinations(b_maj, "LP", raiseException=False)
```

It seems that the core problem comes from the Chord object, because:
```
b_maj = music21.chord.Chord('B D# F#')
print(music21.analysis.neoRiemannian.L(b_maj))
b_maj = music21.chord.Chord([11, 3, 6])
print(music21.analysis.neoRiemannian.L(b_maj))
```
Results in:
```
<music21.chord.Chord A# D# F#>
<music21.chord.Chord B- D# F#>
```

And the `isMinorTriad` method fails on the latter:
```
music21.chord.Chord('B- D# F#').isMinorTriad()
# Returns False
```

**Comment by ax-le** [comment]
The issue seems larger than just fixing chord instanciation for pc sets with numbers, because I've found an example of the same problem, but with symbol instanciation:
```
# Works fine
print(music21.analysis.neoRiemannian.L(music21.chord.Chord([10, 2, 5])))

# Raises an error
print(music21.analysis.neoRiemannian.L(music21.chord.Chord(['A#', 'D', 'F'])))
```

**Comment by jacobtylerwalls** [comment]
Hi -- this should fail, because B- D# F# is not a minor triad. (Strikes some music21 users as unexpected, but the music theory modules care very much about how pitches are spelled, and if a user provides pitch class numbers, they're leaving spelling up to chance. For instance, there is even an `IncorrectlySpelledTriadPrevalence` feature, which would be pointless if music21 didn't care whether triads were spelled correctly.) If your algorithm starts without knowledge of spelling, you should investigate the pitch routines for enharmonic simplification so you can control for spelling before doing LRP transforms.

Thank you for providing an example -- because my question now is why despite `raiseException=False` an exception is being raised. This is the safety check:

https://github.com/cuthbertLab/music21/blob/526bd3cea57aefaa855080d8870aa33bcf9f1ce4/music21/analysis/neoRiemannian.py#L356-L360

But this is what actually fails, and your incorrectly spelled triad falls through the cracks, since incorrectly spelled triads still have a 3-11 Forte class:

https://github.com/cuthbertLab/music21/blob/526bd3cea57aefaa855080d8870aa33bcf9f1ce4/music21/analysis/neoRiemannian.py#L129-L138

Two obvious solutions could be:
 - pass the value of raiseException to the subsequent calls
 - add checks on `isMinorTriad()` and `isMajorTriad()` to the safety check, but this seems to unnecessarily degrade performance for the common case.

**Comment by ax-le** [comment]
Hi,
Thanks for your answer!
I hadn't thought of that fine level of triad spelling (even with the simplifyEnharmonics set to True), but this indeed makes perfect sense now.

As for the `raiseException=False`, I think that passing its value to L, P and R functions should be a better fix, but it could also mislead uncareful users (as the one I was 20 minutes ago) about this triad spelling level, as it would still return a result (but an unexpected one).
What do you think about using the `simplifyEnharmonics` argument both as an output simplifier and an input one?

**Comment by jacobtylerwalls** [comment]
Just to correct my earlier comment, turns out it would be a performance improvement to replace the Forte class check with checks against the two triad types:

```
>>> timeit('c.isMinorTriad()', setup='import music21;c=music21.chord.Chord([60, 66, 70])')
0.29962021499999736
>>> timeit('c.isMajorTriad()', setup='import music21;c=music21.chord.Chord([60, 66, 70])')
0.30119304299999783
>>> timeit('c.forteClassTnI != "3-11"', setup='import music21;c=music21.chord.Chord([60, 66, 70])')
4.618692260000046
```

About uncareful users, I guess I would say that's why `raiseException` defaults True. Is there an example somewhere you were following that led you to set it False?

Regarding running simplifyEnharmonics on input, it would need to be a separate option--because again, people might want spelling preserved on their inputs, but only avoid bizarre spelling on output. IMO we'd be getting into keyword bloat at that point.

However, I think an example showing the utility of running `bestPitches()` would be useful on the Chord class right after it's demo'd that midi numbers are accepted. Something like

>>> c = chord.Chord([58, 62, 65])
>>> c.isTriad()
False

# run bestPitches()
>>> c.isTriad()
True

If we were going to give people some syntactic sugar to do this, I would do it all the way upstream here at the Chord constructor, not in the neoRiemannian module.

Would you like to prepare a patch with one of the two solutions for tightening up the exception handling in the LRP_combinations, include a test, and then add an example of running bestPitches() to the Chord docs? 🤞🏻

**Comment by ax-le** [comment]
Adding a `bestPitches()` example seems to me like a great idea.

> Would you like to prepare a patch with one of the two solutions for tightening up the exception handling in the LRP_combinations, include a test, and then add an example of running bestPitches() to the Chord docs? 🤞🏻

Sure, I'm going to prepare it!
It seems to me that passing `raiseException` is a better fix, as, in my first example, the error was not raised for the first transfrom (L) but for the second one (P):
```
music21.chord.Chord([11, 3, 6]).isMajorTriad() # True
b_maj = music21.chord.Chord([11, 3, 6])
lp_b_maj = music21.analysis.neoRiemannian.LRP_combinations(b_maj, "LP", raiseException=False) # Raises error on the P
```

> turns out it would be a performance improvement to replace the Forte class check with checks against the two triad types

For that matter, if it's ok for you, I will also replace the checks in the same pull request.

**Comment by jacobtylerwalls** [comment]
Oh, that's interesting. The failures can happen in the middle of the chain (if `simplifyEnharmonics=False`). Forgive me if that is what you were referring to when you mentioned users being caught by surprise. Returning one of the intermediate results when raiseException=False (or just silently proceeding to do the rest of transformations!) doesn't seem great -- would it be better to just catch exceptions and then either raise or return the original chord? `raiseException` is not documented on LRP_combinations, but one would assume from the other methods that it returns the original chord if the chain fails.

**Comment by ax-le** [comment]
I actually got confused in my examples sorry.
In
```
>>> music21.analysis.neoRiemannian.LRP_combinations(b_maj, "LP", raiseException=False)
```
The error was raised in the L (because music21.chord.Chord([11, 3, 6]) failed, as stated before).

But in other examples (not previously mentionned), the error is raised in the chain.
For instance in:
```
>>> cs_maj = music21.chord.Chord([1, 5, 8])
>>> music21.analysis.neoRiemannian.LRP_combinations(cs_maj, "LP", simplifyEnharmonics=True)
[...]
LRPException: Cannot perform P on this chord: not a Major or Minor triad
```
Here, the P transform fails because the result of L is not perfectly spelled.
And it happens even if `simplifyEnharmonics=True`, as it only modifies the final output.

That's indeed what I wanted to mention, but I realize that I haven't correctly stated that problem before, my bad!

What is particulary confusing is that the following works
```
>>> cs_maj = music21.chord.Chord("C# E# G#")
>>> music21.analysis.neoRiemannian.LRP_combinations(cs_maj, "LP", simplifyEnharmonics=True)
```

Even if:
```
>>> music21.chord.Chord([1, 5, 8])
music21.chord.Chord("C# E# G#")
```

That's why I wanted to add a `simplifyEnharmonics` kind of argument to simplify the result of each transform, but It may be overkill.
I think that returning the orignal chord or specifying that the chain has failed would be better than returning an intermediate result indeed.

**Comment by jacobtylerwalls** [comment]
Oh definitely -- I I think if `simplifyEnharmonics` is True, it should run on each intermediate result in the chain. No use choking midway if the user is OK with respelling. My word of caution at the outset was just that if we were to perform respelling on the original _input_, then we have to decide whether to edit the chord `inPlace`, and this becomes cumbersome when the issue was just how the chord was originally constructed.

**Comment by jacobtylerwalls** [comment]
Alright, I'm not convinced we've finished investigating this. There shouldn't be a need to respell each intermediate result just to avoid breaking the chain, since the transforms shouldn't output non-triads.

For 

```
>>> cs_maj = music21.chord.Chord([1, 5, 7])
>>> cs_maj
<music21.chord.Chord C# E# G#>
>>> music21.analysis.neoRiemannian.LRP_combinations(cs_maj, "L", simplifyEnharmonics=False)
<music21.chord.Chord C E# G#>
```
I would have expected B# E# G#, because the leading tone in C# is B#. I think if that underlying issue is fixed, then we don't have to add all these intermediate respellings.

**Comment by jacobtylerwalls** [comment]
on this idea, also no need to handle exceptions during the chain, so long as we tighten the safety check to screen out misspelled triads at the beginning.

**Comment by ax-le** [comment]
> My word of caution at the outset was just that if we were to perform respelling on the original input

Indeed, I didn't make this distinction but it has to be made.

>  the transforms shouldn't output non-triads.

Indeed.
I don't know music21 much, but the fact that it works with `music21.chord.Chord("C# E# G#")` and that it doesn't with `music21.chord.Chord([1,5,8])` must be a strong lead, especially because this bug would have been discovered in tests otherwise.

I'm going to look at the Chord constructor and I'll let you know if I find something.

**Comment by ax-le** [comment]
I wasn't successful here, the only interesting thing I found is that:
```
>>> music21.chord.Chord([1,5,8]) == music21.chord.Chord("C# E# G#")
False
```
which was expected due to the behavior of the neoRemannian.L() function, but doesn't seem to be a valid feature.

I may look at it in the next few days but I'm running out of idea right now.

**Comment by jacobtylerwalls** [comment]
The chord constructor is fine--I just think L() needs to transpose by a minor second not an augmented unison to get the new pitch. I can post a small diff in the next couple days.

**Comment by jacobtylerwalls** [comment]
Chord equality checks spelling, so when you leave spelling up to chance there's no expectation it will == any chord with spellings provided.

**Comment by jacobtylerwalls** [comment]
`L()` wasn't the problem; it was transposing by m2 as it should, but we're tripping over this:

https://github.com/cuthbertLab/music21/blob/60d91ddaf4a7398337181a6168c9ec45abcb93b3/music21/pitch.py#L4300-L4310

That's fine, but before the neoRiemannian transforms call transpose(), we should set `thisPitch.spellingIsInferred = False` so that we stay on the rails and always end up with triads.

This diff works:

```diff
diff --git a/music21/analysis/neoRiemannian.py b/music21/analysis/neoRiemannian.py
index a994cc06e..97dacead4 100644
--- a/music21/analysis/neoRiemannian.py
+++ b/music21/analysis/neoRiemannian.py
@@ -184,11 +184,11 @@ def _singlePitchTransform(c, transposeInterval, changingPitch):
     Performs a neoRiemannian transformation on c that involves transposing `changingPitch` by
     `transposeInterval`.
     '''
-    changingPitchCopy = copy.deepcopy(changingPitch)
     newChord = copy.deepcopy(c)
-    for i in range(len(newChord.pitches)):
-        if changingPitchCopy.name == newChord.pitches[i].name:
-            newChord.pitches[i].transpose(transposeInterval, inPlace=True)
+    for p in newChord.pitches:
+        p.spellingIsInferred = False
+        if changingPitch.name == p.name:
+            p.transpose(transposeInterval, inPlace=True)
     return chord.Chord(newChord.pitches)
 ```

```
>>> import music21
>>> cs_maj = music21.chord.Chord([1, 5, 8])
>>> music21.analysis.neoRiemannian.LRP_combinations(cs_maj, "LP", simplifyEnharmonics=False)
<music21.chord.Chord B# E# G##>
>>> music21.analysis.neoRiemannian.LRP_combinations(cs_maj, "LP", simplifyEnharmonics=True)
<music21.chord.Chord C F A>
```

**Comment by jacobtylerwalls** [comment]
Thanks again for the report and for being willing to put together a patch! Sorry I missed that your original input was correctly spelled; I got distracted by the discussion of B- D# F#, which was an intermediate result that you had started to debug and presented separately. 👍🏻

**Comment by ax-le** [comment]
> That's fine, but before the neoRiemannian transforms call transpose(), we should set thisPitch.spellingIsInferred = False so that we stay on the rails and always end up with triads.

Oh, well done! 😃

> All to say -- in the future, please keep the steps to reproduce focused on actually reproducing the error and not a discussion of debugging and proposed fix

You're right! To be honest that's one of my first issues, so this will serve as learning.

I guess that we can close this issue, now that you've opened another one, specific to the core problem?
Thanks for your help!

**Comment by jacobtylerwalls** [comment]
No problem. Again, it was my ruthless skimming that sent us off the rails, apologies!

And, no, I think #929 is separate -- that's inferred being set False when it should stay True; this issue that I targeted in the diff a few comments ago involves making copies of pitches where inferred is True and then never being set False once we start doing transformations that care about strict spelling. So if we fix #929, this problem increases in frequency. :-)

So if you're still up for a patch, we need that diff, some tests, and also the change to test against `.isMinorTriad()` and `.isMajorTriad()` instead of the Forte class, since someone might have incorrectly spelled triads created in some other fashion than using MIDI numbers. It would be a help, but feel no pressure--thanks for your help looking into the issue yesterday!

**Comment by ax-le** [comment]
Oh right, I read the new issue diagonally so I assumed it was the same.

Yes no problem, I'm going to write that patch! :)
Hopefully soon, I'll get back with a Pull Request!
